### PR TITLE
intersphinx: Handle the case where intersphinx_cache_limit is negative

### PR DESCRIPTION
**Pull request by SilverRainZ**
Subject: intersphinx: Handle the case where intersphinx_cache_limit is negative

### Feature or Bugfix

- Bugfix

### Purpose

Ensuring the confval `intersphinx_cache_limit` behaves as documented.

### Detail

The [documentation](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#confval-intersphinx_cache_limit) said:

 > Set this to a negative value to cache inventories
 > for unlimited time.

In the current implementation, a negative `intersphinx_cache_limit` causes
inventories always expire, this patch ensures that it behaves as documented.

### Relates
- https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#confval-intersphinx_cache_limit

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -89,6 +89,9 @@ Bugs fixed
 * #12494: Fix invalid genindex.html file produced with translated docs
   (regression in 7.1.0).
   Patch by Nicolas Peugnet.
+* #12514: intersphinx: Handle the case where ``intersphinx_cache_limit`` is
+  negative, ensure that it behaves as documented.
+  Patch by Shengyu Zhang.
```
Comment (picnixz) [inline CHANGES.rst]:
```suggestion
* #12514: intersphinx: fix the meaning of a negative
  value for :confval:`intersphinx_cache_limit`.
  Patch by Shengyu Zhang.
```

File: sphinx/ext/intersphinx/_load.py [hunk 2]
```diff
@@ -107,15 +107,19 @@ def fetch_inventory_group(
...
     failures = []
     try:
         for inv in invs:
             if not inv:
                 inv = posixpath.join(uri, INVENTORY_FILENAME)
             # decide whether the inventory must be read: always read local
             # files; remote ones only if the cache time is expired
-            if '://' not in inv or uri not in cache or cache[uri][1] < cache_time:
+            if '://' not in inv or uri not in cache or \
+                    (cache_time and cache[uri][1] < cache_time):
```
Comment (picnixz) [inline sphinx/ext/intersphinx/_load.py]:
```suggestion
            if (
            	'://' not in inv or uri not in cache or
                (cache_time and cache[uri][1] < cache_time)
            ):
```

Using line continuations makes it harder to read (and I think we mostly use '()' now).

File: sphinx/ext/intersphinx/_load.py [hunk 3]
```diff
@@ -107,15 +107,19 @@ def fetch_inventory_group(
     app: Sphinx,
     now: int,
 ) -> bool:
-    cache_time = now - app.config.intersphinx_cache_limit * 86400
+    if app.config.intersphinx_cache_limit != -1:
```
Comment (picnixz) [inline sphinx/ext/intersphinx/_load.py]:
```suggestion
    if app.config.intersphinx_cache_limit < 0:
```

The docs says *any* negative value would be fine.

**Review by picnixz — COMMENTED** [review]
Can you add some tests for that as well?

By the way, I'm surprised no one ever complained...

**Comment by SilverRainZ** [comment]
@picnixz I have added a test case, but only test the negative case, I am not sure how to test the positive one for it needs more mocked data.

File: sphinx/ext/intersphinx/_load.py [hunk 4]
```diff
@@ -107,15 +107,21 @@ def fetch_inventory_group(
     app: Sphinx,
     now: int,
 ) -> bool:
-    cache_time = now - app.config.intersphinx_cache_limit * 86400
+    if app.config.intersphinx_cache_limit < 0:
+        cache_time = now - app.config.intersphinx_cache_limit * 86400
+    else:
+        cache_time = None
```
Comment (AA-Turner) [inline sphinx/ext/intersphinx/_load.py]: This would mean that we don't have to change the test below?
```suggestion
        cache_time = 0
```

Comment (SilverRainZ) [reply sphinx/ext/intersphinx/_load.py:113]: Yes, but the semantics don't seem very clear?

Comment (SilverRainZ) [reply sphinx/ext/intersphinx/_load.py]: fixed.

**Comment by picnixz** [comment]
> I am not sure how to test the positive one for it needs more mocked data.

If I have time tomorrow (I've got some reviews to do + some stuff to complete on CPython), I'll make a PR on your forked branch (or just dump the code here). But I'm not sure I'll have time. Otherwise, I think it's fine. I just need to remember how I mocked my own tests for those kind of things...

**Review by firefly9966 — APPROVED** [review]

**Comment by SilverRainZ** [comment]
rebased to master again.

**Comment by SilverRainZ** [comment]
I can not run my test because of the following error:

```
\ $ python -m pytest -v tests/test_extensions/test_ext_intersphinx.py::test_intersphinx_cache_limit
ImportError while loading conftest '/Users/bytedance/git/public/sphinx-la/tests/conftest.py'.
tests/conftest.py:15: in <module>
    from sphinx.testing.util import _clean_up_global_state
sphinx/testing/util.py:17: in <module>
    import sphinx.application
sphinx/application.py:30: in <module>
    from sphinx.highlighting import lexer_classes
sphinx/highlighting.py:86: in <module>
    class PygmentsBridge:
sphinx/highlighting.py:89: in PygmentsBridge
    html_formatter = HtmlFormatter[str]
E   TypeError: type 'HtmlFormatter' is not subscriptable
```

**Comment by picnixz** [comment]
You should update your version of pygments to include this commit: https://github.com/pygments/pygments/commit/0f4061447fd583da51eed2cb10a4a2813d00ac7e

**Comment by AA-Turner** [comment]
> > I am not sure how to test the positive one for it needs more mocked data.
> 
> If I have time tomorrow (I've got some reviews to do + some stuff to complete on CPython), I'll make a PR on your forked branch (or just dump the code here). But I'm not sure I'll have time. Otherwise, I think it's fine. I just need to remember how I mocked my own tests for those kind of things...

@picnixz did you want to add these tests or are we OK to merge?

A

**Comment by picnixz** [comment]
> did you want to add these tests or are we OK to merge?

I shamefully forgot. I'm fine with your test (much smaller, less complicate).

**Comment by AA-Turner** [comment]
> I shamefully forgot. I'm fine with your test (much smaller, less complicate).

There's no shame in being busy! Thanks for the quick reply

A

**Pull request merged**